### PR TITLE
Material Serialization

### PR DESCRIPTION
**Pull request by jtramm**
This PR serializes materials into global arrays. The global arrays are implemented in the form of a new class, `vector2d<T>`. See the writeup of PR #8 for more info.

### CODE REVIEW

File: include/openmc/material.h [hunk 1]
```diff
@@ -170,6 +183,7 @@ class Material
   vector<ThermalTable> thermal_tables_;
 
   Bremsstrahlung ttb_;
+  gsl::index index_;
```
Comment (paulromano) [inline include/openmc/material.h]: Looks to me like this could still be private?

File: include/openmc/vector.h [hunk 2]
```diff
@@ -137,6 +138,8 @@ class vector {
     size_ = count;
   }
 
+  size_type footprint() { return size_ * sizeof(T); }
```
Comment (paulromano) [inline include/openmc/vector.h]: Consider renaming `nbytes` to be more explicit

File: include/openmc/vector.h [hunk 3]
```diff
@@ -388,6 +391,77 @@ void vector<T>::assign(InputIt first, InputIt last)
...
+template<typename T>
+class vector2d : public vector<T> {
+  public:
+  using size_type = std::size_t;
+  using const_iterator = const T*;
+  using reference = T&;
+  using const_reference = const T&;
+  using vector<T>::data_;
+  using vector<T>::capacity_;
+  using vector<T>::size_;
```
Comment (paulromano) [inline include/openmc/vector.h]: This makes these three data members public -- is that really necessary?

File: src/device_alloc.cpp [hunk 4]
```diff
@@ -184,33 +184,99 @@ void move_read_only_data_to_device()
...
+    std::cout << " Fissionable Material Statistics:" << std::endl <<
+      "   Max Nuclide Count: " << max << std::endl <<
+      "   Min Nuclide Count: " << min << std::endl <<
+      "   # Fissionable Materials with >= 200 Nuclides: " << n_over_200 << std::endl <<
+      "   # Fissionable Materials with  < 200 Nuclides: " << n_under_200 << std::endl;
+  }
+
+  // Determine size of inner dimension for serialized material vectors
+  for (int i = 0; i < model::materials_size; i++) {
+    auto& mat = model::materials[i];
```
Comment (paulromano) [inline src/device_alloc.cpp]:
```suggestion
    const auto& mat = model::materials[i];
```

File: include/openmc/vector.h [hunk 5]
```diff
@@ -388,6 +391,77 @@ void vector<T>::assign(InputIt first, InputIt last)
...
+      // If new size is smaller, call destructor on erased elements
+      for (size_type i = count; i < size_; ++i) {
+        data_[i].~T();
+      }
+    }
+    size_ = count;
+  }
+
+  // Copies a row of data into the 2D matrix at row i
+  void copy_row(size_type i, vector<T>& row) {
```
Comment (paulromano) [inline include/openmc/vector.h]:
```suggestion
  void copy_row(size_type i, const vector<T>& row) {
```

File: include/openmc/vector.h [hunk 6]
```diff
@@ -388,6 +391,77 @@ void vector<T>::assign(InputIt first, InputIt last)
...
+    for (int j = 0; j < row.size(); j++) {
+      data_[i * stride_ + j] = row[j];
+    }
+  }
+  void copy_row(size_type i, xt::xtensor<T,1>& row) {
```
Comment (paulromano) [inline include/openmc/vector.h]:
```suggestion
  void copy_row(size_type i, const xt::xtensor<T,1>& row) {
```

File: include/openmc/vector.h [hunk 7]
```diff
@@ -388,6 +391,77 @@ void vector<T>::assign(InputIt first, InputIt last)
...
+  const_reference operator()(size_type outer_pos, size_type pos) const { return data_[outer_pos * stride_ + pos]; }
+
+  // Stretching functions.
+  // The idea here is that you should call these on every row you want to add to the
+  // matrix first to determine which is longest, which is kept track of in the stride_
+  // field. We could also add a direct way of setting stride_
+  // down the line, but these stretching functions are useful for finding the material
+  // with the most nuclides, thermal tables, etc, without having to repeat the logic
+  // multiple times elsewhere.
+  void stretch(vector<T>& vect) {
```
Comment (paulromano) [inline include/openmc/vector.h]:
```suggestion
  void stretch(const vector<T>& vect) {
```

File: include/openmc/vector.h [hunk 8]
```diff
@@ -388,6 +391,77 @@ void vector<T>::assign(InputIt first, InputIt last)
...
+    if (vect.size() > stride_) {
+      stride_ = vect.size();
+    }
+  }
+  void stretch(xt::xtensor<T,1>& vect) {
```
Comment (paulromano) [inline include/openmc/vector.h]:
```suggestion
  void stretch(const xt::xtensor<T,1>& vect) {
```

**Review by paulromano — CHANGES_REQUESTED** [review]
This one looks fairly straightforward -- thanks! Just a few small comments.

What performance improvement did you see from just this optimization?

File: include/openmc/material.h (hunk 1 above)
Comment (jtramm) [reply include/openmc/material.h]: Good eyes! Yes, in a previous version I was accessing the `index_` field directly, but changed it to use the accessors instead so this should have been restored to private.

**Comment by jtramm** [comment]
Performance delta on the A100 for this PR is as follows:

  | Old | This PR
-- | -- | --
HM Large Inactive | 356 | 361
SMR Active | 161 | 160

However, it is notable that the differences can be much more dramatic on other GPU architectures (and perhaps even between compilers for A100 down the line) due to how various OpenMP runtimes handle the mapping of millions of pointers.

**Review by paulromano — APPROVED** [review]

**Pull request merged**